File: python/02_producer_consumer_queue.py
"""
PROBLEM: Producer-Consumer Queue

Implement a thread-safe queue where producer threads add items and consumer threads
process them. This is a classic concurrent programming pattern.

REQUIREMENTS:
- Implement a bounded queue with a maximum size
- Producers should block when queue is full
- Consumers should block when queue is empty
- Must handle proper signaling between producers and consumers
- Should use threading.Condition or similar synchronization primitives

PERFORMANCE NOTES:
- Should handle 100+ items per second throughput
- Producers and consumers should not busy-wait
- Should minimize lock contention

TEST CASE EXPECTATIONS:
- Single producer adding 100 items, single consumer processing all should complete correctly
- Multiple producers and consumers should maintain correct ordering/counts
- Queue of size 5 with 3 producers and 3 consumers should not overflow
"""

import threading
import time
from typing import Any, Optional


class ThreadSafeQueue:
    """A thread-safe queue for producer-consumer patterns."""

    def __init__(self, max_size: int = 10):
        """
        Initialize the queue.

        Args:
            max_size: Maximum number of items the queue can hold
        """
        self.max_size = max_size
        # TODO: Initialize queue, locks, and condition variables
        self.items = []

    def put(self, item: Any, timeout: Optional[float] = None) -> bool:
        """
        Add an item to the queue. Blocks if queue is full.

        Args:
            item: Item to add to the queue
            timeout: Maximum time to wait (None = wait forever)

        Returns:
            True if item was added, False if timeout occurred
        """
        # TODO: Implement with proper synchronization
        # Block if queue is full
        # Signal waiting consumers that an item is available
        pass

    def get(self, timeout: Optional[float] = None) -> tuple[bool, Any]:
        """
        Remove and return an item from the queue. Blocks if queue is empty.

        Args:
            timeout: Maximum time to wait (None = wait forever)

        Returns:
            Tuple of (success, item) - success is False if timeout occurred
        """
        # TODO: Implement with proper synchronization
        # Block if queue is empty
        # Signal waiting producers that space is available
        pass

    def size(self) -> int:
        """Get current number of items in queue."""
        # TODO: Return size safely
        pass

    def is_empty(self) -> bool:
        """Check if queue is empty."""
        # TODO: Implement thread-safely
        pass

    def is_full(self) -> bool:
        """Check if queue is full."""
        # TODO: Implement thread-safely
        pass


def test_single_producer_consumer():
    """Test single producer and consumer."""
    queue = ThreadSafeQueue(max_size=5)
    produced = []
    consumed = []

    def producer():
        for i in range(20):
            queue.put(i)
            produced.append(i)
            time.sleep(0.01)

    def consumer():
        for _ in range(20):
            success, item = queue.get()
            if success:
                consumed.append(item)
                time.sleep(0.015)

    prod_thread = threading.Thread(target=producer)
    cons_thread = threading.Thread(target=consumer)

    prod_thread.start()
    cons_thread.start()

    prod_thread.join()
    cons_thread.join()

    assert produced == consumed, f"Mismatch: produced {produced}, consumed {consumed}"
    print(f"✓ Single producer-consumer test passed ({len(consumed)} items)")


def test_multiple_producers_consumers():
    """Test multiple producers and consumers."""
    queue = ThreadSafeQueue(max_size=10)
    num_producers = 3
    num_consumers = 2
    items_per_producer = 30
    all_produced = set()
    all_consumed = []
    lock = threading.Lock()

    def producer(producer_id):
        for i in range(items_per_producer):
            item = f"producer_{producer_id}_item_{i}"
            queue.put(item)
            with lock:
                all_produced.add(item)

    def consumer(consumer_id):
        for _ in range(items_per_producer):
            success, item = queue.get()
            if success:
                with lock:
                    all_consumed.append(item)

    # Create and start threads
    producers = [
        threading.Thread(target=producer, args=(i,)) for i in range(num_producers)
    ]
    consumers = [
        threading.Thread(target=consumer, args=(i,)) for i in range(num_consumers)
    ]

    for thread in producers:
        thread.start()

    for thread in consumers:
        thread.start()

    for thread in producers:
        thread.join()

    for thread in consumers:
        thread.join()

    assert len(all_produced) == num_producers * items_per_producer
    assert len(all_consumed) == num_producers * items_per_producer
    assert set(all_consumed) == all_produced

    print(f"✓ Multiple producers-consumers test passed ({len(all_consumed)} items)")


def test_queue_blocking():
    """Test that queue properly blocks when full/empty."""
    queue = ThreadSafeQueue(max_size=3)

    # Fill the queue
    for i in range(3):
        success = queue.put(i, timeout=1.0)
        assert success, "Should be able to put items in empty queue"

    # Try to put with timeout (should timeout since queue is full)
    success = queue.put(99, timeout=0.5)
    assert not success, "Should timeout when queue is full"

    # Get an item to make space
    success, item = queue.get(timeout=1.0)
    assert success and item == 0, "Should be able to get from non-empty queue"

    # Now we should be able to put
    success = queue.put(99, timeout=1.0)
    assert success, "Should be able to put after making space"

    print("✓ Queue blocking test passed")


def test_queue_empty_get():
    """Test that get blocks/times out on empty queue."""
    queue = ThreadSafeQueue(max_size=5)

    success, item = queue.get(timeout=0.5)
    assert not success, "Should timeout on empty queue"

    print("✓ Empty queue get test passed")


if __name__ == "__main__":
    test_single_producer_consumer()
    test_multiple_producers_consumers()
    test_queue_blocking()
    test_queue_empty_get()
    print("\n✓ All tests passed!")

File: cpp/02_producer_consumer_queue.cpp
/*
PROBLEM: Producer-Consumer Queue

Implement a thread-safe queue where producer threads add items and consumer threads
remove them. This classic concurrent pattern requires careful synchronization.

REQUIREMENTS:
- Implement a template-based bounded queue
- Producers block when queue is full
- Consumers block when queue is empty
- Use std::condition_variable for efficient waiting
- Use std::mutex for synchronization
- RAII patterns throughout

PERFORMANCE NOTES:
- Should handle 100+ items per second throughput
- Minimal busy-waiting (use condition variables)
- Lock contention should be minimized

TEST CASE EXPECTATIONS:
- Single producer adding 1000 items, single consumer getting all should work correctly
- Multiple producers and consumers should maintain integrity
- Queue of size 10 with 5 producers and 5 consumers should not overflow
- notify_one() should efficiently wake waiting threads
*/

#include <iostream>
#include <queue>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <vector>
#include <cassert>
#include <chrono>
#include <optional>

template <typename T>
class ThreadSafeQueue {
public:
    ThreadSafeQueue(size_t max_size = 10) : max_size_(max_size) {
        // TODO: Initialize queue, mutex, and condition variables
    }

    // Add an item to queue, block if full
    // Returns true if added, false if timeout
    bool put(const T& item, double timeout_seconds = -1.0) {
        // TODO: Implement with proper synchronization
        // - Acquire lock
        // - Wait if queue is full (using condition_variable with timeout)
        // - Add item
        // - Notify waiting consumers
        // - Return success/timeout status
        return false;
    }

    // Get an item from queue, block if empty
    // Returns {success, item} - success is false if timeout
    std::pair<bool, std::optional<T>> get(double timeout_seconds = -1.0) {
        // TODO: Implement with proper synchronization
        // - Acquire lock
        // - Wait if queue is empty (using condition_variable with timeout)
        // - Remove and return item
        // - Notify waiting producers
        // - Return success and item
        return {false, std::nullopt};
    }

    size_t size() const {
        // TODO: Return size thread-safely
        return 0;
    }

    bool is_empty() const {
        // TODO: Implement thread-safely
        return true;
    }

    bool is_full() const {
        // TODO: Implement thread-safely
        return false;
    }

private:
    std::queue<T> queue_;
    size_t max_size_;
    // TODO: Add std::mutex and std::condition_variable
};


void test_single_producer_consumer() {
    ThreadSafeQueue<int> queue(5);
    std::vector<int> produced;
    std::vector<int> consumed;

    auto producer = [&queue, &produced]() {
        for (int i = 0; i < 100; ++i) {
            queue.put(i);
            produced.push_back(i);
        }
    };

    auto consumer = [&queue, &consumed]() {
        for (int i = 0; i < 100; ++i) {
            auto [success, item] = queue.get();
            if (success && item.has_value()) {
                consumed.push_back(item.value());
            }
        }
    };

    std::thread prod_thread(producer);
    std::thread cons_thread(consumer);

    prod_thread.join();
    cons_thread.join();

    assert(produced == consumed);
    std::cout << "✓ Single producer-consumer test passed (" << consumed.size() << " items)\n";
}


void test_multiple_producers_consumers() {
    ThreadSafeQueue<int> queue(10);
    const int num_producers = 3;
    const int num_consumers = 3;
    const int items_per_producer = 50;
    std::mutex stats_lock;
    int total_produced = 0;
    int total_consumed = 0;

    auto producer = [&queue, &stats_lock, &total_produced](int producer_id) {
        for (int i = 0; i < 50; ++i) {
            queue.put(producer_id * 1000 + i);
            {
                std::lock_guard<std::mutex> lock(stats_lock);
                total_produced++;
            }
        }
    };

    auto consumer = [&queue, &stats_lock, &total_consumed]() {
        while (true) {
            auto [success, item] = queue.get(0.1); // 100ms timeout
            if (success && item.has_value()) {
                {
                    std::lock_guard<std::mutex> lock(stats_lock);
                    total_consumed++;
                }
            } else {
                // Small timeout reached, check if we should continue
                // In a real scenario, you'd have a shutdown signal
                break;
            }
        }
    };

    std::vector<std::thread> threads;

    // Create producers
    for (int i = 0; i < num_producers; ++i) {
        threads.emplace_back(producer, i);
    }

    // Create consumers
    for (int i = 0; i < num_consumers; ++i) {
        threads.emplace_back(consumer);
    }

    for (auto& thread : threads) {
        thread.join();
    }

    // Wait a bit for remaining items to be consumed
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    int expected = num_producers * items_per_producer;
    // Note: This test might have race conditions in counting, adjust as needed
    std::cout << "✓ Multiple producers-consumers test passed (" << total_consumed << " items consumed)\n";
}


void test_queue_blocking() {
    ThreadSafeQueue<int> queue(3);

    // Fill the queue
    for (int i = 0; i < 3; ++i) {
        bool success = queue.put(i);
        assert(success);
    }

    // Try to put with timeout (should timeout since queue is full)
    bool success = queue.put(99, 0.1); // 100ms timeout
    assert(!success); // Should timeout

    // Get an item to make space
    auto [get_success, item] = queue.get();
    assert(get_success && item.has_value());
    assert(item.value() == 0);

    // Now we should be able to put
    success = queue.put(99);
    assert(success);

    std::cout << "✓ Queue blocking test passed\n";
}


void test_queue_empty_get() {
    ThreadSafeQueue<int> queue(5);

    auto [success, item] = queue.get(0.1); // 100ms timeout
    assert(!success); // Should timeout on empty queue

    std::cout << "✓ Empty queue get test passed\n";
}


void benchmark_throughput() {
    ThreadSafeQueue<int> queue(100);
    const int num_items = 10000;

    auto start = std::chrono::high_resolution_clock::now();

    std::thread producer([&queue, num_items]() {
        for (int i = 0; i < num_items; ++i) {
            queue.put(i);
        }
    });

    std::thread consumer([&queue, num_items]() {
        for (int i = 0; i < num_items; ++i) {
            queue.get();
        }
    });

    producer.join();
    consumer.join();

    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);

    double items_per_second = (num_items * 1000.0) / duration.count();
    std::cout << "✓ Throughput benchmark: " << items_per_second << " items/second\n";
}


int main() {
    std::cout << "Running ThreadSafeQueue tests...\n\n";

    test_single_producer_consumer();
    test_multiple_producers_consumers();
    test_queue_blocking();
    test_queue_empty_get();
    benchmark_throughput();

    std::cout << "\n✓ All tests passed!\n";

    return 0;
}
